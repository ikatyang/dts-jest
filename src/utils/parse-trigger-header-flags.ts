import { TriggerHeaderFlags } from '../definitions';

export const parse_trigger_header_flags = (unparsed_flags: string) => {
  let flags: TriggerHeaderFlags = 0;

  unparsed_flags
    .split(/(?=:)/)
    .filter(flag_literal => flag_literal.length !== 0)
    .forEach(flag_literal => {
      switch (flag_literal) {
        case TriggerHeaderFlags[TriggerHeaderFlags[':fail']]:
        case TriggerHeaderFlags[TriggerHeaderFlags[':pass']]:
        case TriggerHeaderFlags[TriggerHeaderFlags[':show']]:
        case TriggerHeaderFlags[TriggerHeaderFlags[':snap']]:
        case TriggerHeaderFlags[TriggerHeaderFlags[':not-any']]:
        case TriggerHeaderFlags[TriggerHeaderFlags[':only']]:
        case TriggerHeaderFlags[TriggerHeaderFlags[':skip']]:
        case TriggerHeaderFlags[TriggerHeaderFlags[':group']]:
          const current_flag: TriggerHeaderFlags = TriggerHeaderFlags[
            flag_literal as any
          ] as any;
          if (flags & current_flag) {
            throw new Error(`Duplicate flag '${flag_literal}'`);
          }
          flags = flags | current_flag;
          break;
        default:
          throw new Error(`Unexpected flag '${flag_literal}'`);
      }
    });

  if (
    flags & TriggerHeaderFlags[':pass'] &&
    flags & TriggerHeaderFlags[':fail']
  ) {
    const pass_flag_literal = TriggerHeaderFlags[TriggerHeaderFlags[':pass']];
    const fail_flag_literal = TriggerHeaderFlags[TriggerHeaderFlags[':fail']];
    throw new Error(
      `${pass_flag_literal} and ${fail_flag_literal} cannot be used simultaneously'`,
    );
  }

  if (
    flags & TriggerHeaderFlags[':group'] &&
    flags & TriggerHeaderFlags.Assertion
  ) {
    const group_flag_literal = TriggerHeaderFlags[TriggerHeaderFlags[':group']];
    throw new Error(
      `AssertionFlag is not allowed to be used with flag '${group_flag_literal}'`,
    );
  }

  return flags;
};
