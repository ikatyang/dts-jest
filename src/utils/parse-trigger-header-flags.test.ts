import { parse_trigger_header_flags } from './parse-trigger-header-flags';

it('should parse correclty', () => {
  expect(parse_trigger_header_flags(':pass:snap')).toMatchSnapshot();
});

it('should throw error if there is a duplicate flag', () => {
  expect(() =>
    parse_trigger_header_flags(':fail:fail'),
  ).toThrowErrorMatchingSnapshot();
});

it('should throw error if there is an unexpected flag', () => {
  expect(() =>
    parse_trigger_header_flags(':unexpected'),
  ).toThrowErrorMatchingSnapshot();
});

it('should throw error if group flag and assertion flag appear simultaneously ', () => {
  expect(() =>
    parse_trigger_header_flags(':group:pass'),
  ).toThrowErrorMatchingSnapshot();
});

it('should throw error if pass flag and fail flag appear simultaneously ', () => {
  expect(() =>
    parse_trigger_header_flags(':pass:fail'),
  ).toThrowErrorMatchingSnapshot();
});
