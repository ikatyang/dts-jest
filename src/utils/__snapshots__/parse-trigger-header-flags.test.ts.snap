// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should parse correclty 1`] = `5`;

exports[`should throw error if group flag and assertion flag appear simultaneously  1`] = `"AssertionFlag is not allowed to be used with flag ':group'"`;

exports[`should throw error if pass flag and fail flag appear simultaneously  1`] = `":pass and :fail cannot be used simultaneously'"`;

exports[`should throw error if there is a duplicate flag 1`] = `"Duplicate flag ':fail'"`;

exports[`should throw error if there is an unexpected flag 1`] = `"Unexpected flag ':unexpected'"`;
